mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td

```diff
@@ -1428,7 +1428,12 @@ def LLVM_LLVMFuncOp : LLVM_Op<"func", [
     OptionalAttr<LLVM_VScaleRangeAttr>:$vscale_range,
     OptionalAttr<FramePointerKindAttr>:$frame_pointer,
     OptionalAttr<StrAttr>:$target_cpu,
-    OptionalAttr<LLVM_TargetFeaturesAttr>:$target_features
+    OptionalAttr<LLVM_TargetFeaturesAttr>:$target_features,
+    OptionalAttr<BoolAttr>:$unsafe_fp_math,
+    OptionalAttr<BoolAttr>:$no_infs_fp_math,
+    OptionalAttr<BoolAttr>:$no_nans_fp_math,
+    OptionalAttr<BoolAttr>:$approx_func_fp_math,
+    OptionalAttr<BoolAttr>:$no_signed_zeros_fp_math
```

Comment:
I am sadly not too familier with these attributes, and documentation seems very sparse, so I am curious whether these attributes are best modelled as `OptionalAttr<BoolAttr>` or `OptionalAttr<UnitAttr>`.

The former is tri-state between "not present", "true" and "false", but does LLVM actually differentiate between "not-present" and "false" anywhere? If not, making it a `OptionalAttr<UnitAttr>` would probably be cleaner.

Reply:
That's a great question. As you say, documentation is unfortunately very sparse. I decided that closely matching the representation of the LLVM IR attribute may be less confusing and less prone to error, but I could be wrong.

It looks like the case you'd want to have an explicit false is for interaction with the `-enable-unsafe-fp-math` flag. `llvm/test/CodeGen/X86/change-unsafe-fp-math.ll` appears to test this interaction, where the `"unsafe-fp-math="false"` attribute does indeed override the codegen flag. So my conclusion is that BoolAttr is probably needed.